### PR TITLE
[FLINK-9637] Add public user documentation for state TTL feature

### PR DESCRIPTION
**Pull request by azagrebin**
## What is the purpose of the change

Public user documentation for first iteration of state TTL feature

## Brief change log

add TTL section in docs/dev/stream/state/state.md

## Verifying this change

check TTL doc section

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: docs/dev/stream/state/state.md [hunk 1]
```diff
@@ -266,6 +266,92 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
 we start over from `0`. Note that this would keep a different state value for each different input
 key if we had tuples with different values in the first field.
 
+### State time-to-live (TTL)
+
+A time-to-live (TTL) can be assigned to the keyed state value. 
+In this case it will expire after the configured TTL
+and its stored value will be cleaned up based on the best effort.
```
Comment (eliaslevy) [inline docs/dev/stream/state/state.md]: s/based on the best effort/on a best effort basis/

File: docs/dev/stream/state/state.md [hunk 2]
```diff
@@ -266,6 +266,92 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
...
+Depending on configuration, the expired state can become unavailable for read access
+even if it is not cleaned up yet. In this case it behaves as if it does not exist any more.
```
Comment (eliaslevy) [inline docs/dev/stream/state/state.md]: s/even if it is not cleaned up yet/even if it has yet to be removed/
s/In this case it behaves as if it does not exist any more/In that case, it behaves as if it no longer exists/

File: docs/dev/stream/state/state.md [hunk 3]
```diff
@@ -266,6 +266,92 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
...
+
+The collection types of state support TTL on entry level: 
+separate list elements and map entries expire independently.
```
Comment (eliaslevy) [inline docs/dev/stream/state/state.md]: The state collection types support per-entry TTLs: list elements and map entries expire independently.

File: docs/dev/stream/state/state.md [hunk 4]
```diff
@@ -266,6 +266,92 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
...
+A time-to-live (TTL) can be assigned to the keyed state value. 
+In this case it will expire after the configured TTL
+and its stored value will be cleaned up based on the best effort.
+Depending on configuration, the expired state can become unavailable for read access
+even if it is not cleaned up yet. In this case it behaves as if it does not exist any more.
+
+The collection types of state support TTL on entry level: 
+separate list elements and map entries expire independently. 
+
+The behaviour of state with TTL firstly should be configured by building `StateTtlConfiguration`:
```
Comment (eliaslevy) [inline docs/dev/stream/state/state.md]: To use state TTL you must first build a `StateTtlConfiguration` object:

File: docs/dev/stream/state/state.md [hunk 5]
```diff
@@ -266,6 +266,92 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
...
+    .newBuilder(Time.seconds(1))
+    .setTtlUpdateType(StateTtlConfiguration.TtlUpdateType.OnCreateAndWrite)
+    .setStateVisibility(StateTtlConfiguration.TtlStateVisibility.NeverReturnExpired)
+    .build()
+{% endhighlight %}
+</div>
+</div>
+
+It has several options to consider. 
+The first parameter of `newBuilder` method is mandatory, it is a value of time-to-live itself.
```
Comment (eliaslevy) [inline docs/dev/stream/state/state.md]: The first parameter of `newBuilder` method is mandatory, it is the time-to-live value.

File: docs/dev/stream/state/state.md [hunk 6]
```diff
@@ -266,6 +266,92 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
...
+
+The update type configures when the time-to-live of state value is prolonged (default `OnCreateAndWrite`):
```
Comment (eliaslevy) [inline docs/dev/stream/state/state.md]: The update type configures when the state TTL is refreshed (default `OnCreateAndWrite`):

File: docs/dev/stream/state/state.md [hunk 7]
```diff
@@ -266,6 +266,92 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
...
+ - `StateTtlConfiguration.TtlUpdateType.OnCreateAndWrite` - only on creation and write access,
+ - `StateTtlConfiguration.TtlUpdateType.OnReadAndWrite` - also on read access.
+ 
+The state visibility configures whether the expired value is returned on read access 
+if it is not cleaned up yet (default `NeverReturnExpired`):
+
+ - `StateTtlConfiguration.TtlStateVisibility.NeverReturnExpired` - expired value is never returned,
+ - `StateTtlConfiguration.TtlStateVisibility.ReturnExpiredIfNotCleanedUp` - returned if still available.
+
+The TTL can be enabled in descriptor for any type of state:
```
Comment (eliaslevy) [inline docs/dev/stream/state/state.md]: TTL functionality can be enabled in the descriptor of any type of state:

File: docs/dev/stream/state/state.md [hunk 8]
```diff
@@ -266,6 +266,92 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
...
+</div>
+
+**Notes:** 
+
+- The state backends store the timestamp of last modification along with the user value, 
+which means that enabling this feature increases consumption of state storage.
+
+- As of current implementation the state storage is cleaned up of expired value 
+only on its explicit read access per key, e.g. calling `ValueState.value()`. 
+This might change in future releases, e.g. additional strategies might be added in background to speed up cleanup.
```
Comment (eliaslevy) [inline docs/dev/stream/state/state.md]: In the current implementation, expired values are only removed when they are read explicitly, e.g. by calling `ValueState.value()`. Note that this means that under the current implementation if expired state is not read, it won't be removed, possibly leading to ever growing state. This might change in future releases. Additional strategies might be added that clean up expired state in the background. 

BTW, is this true given FLINK-9938?

File: docs/dev/stream/state/state.md (hunk 2 above)
Comment (StefanRRichter) [reply docs/dev/stream/state/state.md]: This is very abstract in isolation, maybe we should discuss this in the context of the different visibility levels to make it more clear that this is what we are talking about here?

File: docs/dev/stream/state/state.md (hunk 8 above)
Comment (StefanRRichter) [reply docs/dev/stream/state/state.md]: I think if we merge FLINK-9938 before release, the description should be updated.

File: docs/dev/stream/state/state.md (hunk 1 above)
Comment (StefanRRichter) [inline docs/dev/stream/state/state.md]: I was expecting some additional info about what "best effort" means here, but it came a bit later. Maybe we should say that this is discussed in more detail later. I think it might also make sense to name some use cases that are currently supported (and some that are not/will be in the future) to make things a bit more concrete for users.

File: docs/dev/stream/state/state.md [hunk 9]
```diff
@@ -266,6 +266,92 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
...
+
+- Only *processing time* scale is currently supported for TTL.
+
+- Trying to restore state, which was previously configured without TTL, using TTL enabled descriptor or vice versa
+will lead to compatibility failure.
```
Comment (StefanRRichter) [inline docs/dev/stream/state/state.md]: maybe name the concrete exception

File: docs/dev/stream/state/state.md [hunk 10]
```diff
@@ -266,6 +266,92 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
...
+separate list elements and map entries expire independently. 
+
+The behaviour of state with TTL firstly should be configured by building `StateTtlConfiguration`:
+
+<div class="codetabs" markdown="1">
+<div data-lang="java" markdown="1">
+{% highlight java %}
+StateTtlConfiguration ttlConfig = StateTtlConfiguration
+    .newBuilder(Time.seconds(1))
+    .setTtlUpdateType(StateTtlConfiguration.TtlUpdateType.OnCreateAndWrite)
```
Comment (bowenli86) [inline docs/dev/stream/state/state.md]: My gut feeling is that the class names are quite verbose currently - like duplicated 'Ttl' in the namings. If it's not too late for release 1.6.0, would be nice to shorten it to something like `StateTtlConfig.UpdateType.OnCreateAndWrite`

File: docs/dev/stream/state/state.md (hunk 8 above)
Comment (azagrebin) [reply docs/dev/stream/state/state.md]: I think we still want to improve this, because FLINK-9938 is limited in a way that it cleans up only the snapshot data but not the local one used by running pipeline atm.

File: docs/dev/stream/state/state.md (hunk 10 above)
Comment (azagrebin) [reply docs/dev/stream/state/state.md]: Agree, I added the renaming commit in PR #6460 for FLINK-9938.

**Comment by azagrebin** [comment]
Thanks for review guys! I updated the PR.

File: docs/dev/stream/state/state.md [hunk 11]
```diff
@@ -266,6 +266,101 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
 we start over from `0`. Note that this would keep a different state value for each different input
 key if we had tuples with different values in the first field.
 
+### State time-to-live (TTL)
+
+A time-to-live (TTL) can be assigned to the keyed state value.
```
Comment (twalthr) [inline docs/dev/stream/state/state.md]: `to each keyed state value`

File: docs/dev/stream/state/state.md [hunk 12]
```diff
@@ -266,6 +266,101 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
...
+and its stored value will be cleaned up on the best effort basis which is discussed in details later.
+
+The state collection types support per-entry TTLs: list elements and map entries expire independently.
+
+To use state TTL you must first build a `StateTtlConfiguration` object:
+
+<div class="codetabs" markdown="1">
+<div data-lang="java" markdown="1">
+{% highlight java %}
+StateTtlConfiguration ttlConfig = StateTtlConfiguration
```
Comment (twalthr) [inline docs/dev/stream/state/state.md]: Include imports to find the newly introduced class. Also to clarify which Flink `Time` class we are using.

File: docs/dev/stream/state/state.md [hunk 13]
```diff
@@ -266,6 +266,101 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
...
+val ttlConfig = StateTtlConfiguration
+    .newBuilder(Time.seconds(1))
+    .setTtlUpdateType(StateTtlConfiguration.TtlUpdateType.OnCreateAndWrite)
+    .setStateVisibility(StateTtlConfiguration.TtlStateVisibility.NeverReturnExpired)
+    .build()
+{% endhighlight %}
+</div>
+</div>
+
+It has several options to consider.
```
Comment (twalthr) [inline docs/dev/stream/state/state.md]: What is "it"?

File: docs/dev/stream/state/state.md [hunk 14]
```diff
@@ -266,6 +266,101 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
...
+The update type configures when the state TTL is refreshed (default `OnCreateAndWrite`):
+
+ - `StateTtlConfiguration.TtlUpdateType.OnCreateAndWrite` - only on creation and write access,
+ - `StateTtlConfiguration.TtlUpdateType.OnReadAndWrite` - also on read access.
+ 
+The state visibility configures whether the expired value is returned on read access 
+if it is not cleaned up yet (default `NeverReturnExpired`):
+
+ - `StateTtlConfiguration.TtlStateVisibility.NeverReturnExpired` - expired value is never returned,
+ - `StateTtlConfiguration.TtlStateVisibility.ReturnExpiredIfNotCleanedUp` - returned if still available.
```
Comment (twalthr) [inline docs/dev/stream/state/state.md]: IMHO we could have shorted the names a little bit and moved `TtlStateVisibility` to a separate class. But now it is too late.

File: docs/dev/stream/state/state.md [hunk 15]
```diff
@@ -266,6 +266,101 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
...
+ 
+ In case of `NeverReturnExpired`, the expired state behaves as if it does not exist any more,
```
Comment (twalthr) [inline docs/dev/stream/state/state.md]: `anymore`?

File: docs/dev/stream/state/state.md [hunk 16]
```diff
@@ -266,6 +266,101 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
...
+
+A time-to-live (TTL) can be assigned to the keyed state value. 
+In this case it will expire after the configured TTL
+and its stored value will be cleaned up on the best effort basis which is discussed in details later.
+
+The state collection types support per-entry TTLs: list elements and map entries expire independently.
+
+To use state TTL you must first build a `StateTtlConfiguration` object:
+
+<div class="codetabs" markdown="1">
```
Comment (twalthr) [inline docs/dev/stream/state/state.md]: Show in the first example how this feature is used. Not only the configuration without context. Maybe with a example state for which the config is set.

File: docs/dev/stream/state/state.md [hunk 17]
```diff
@@ -266,6 +266,101 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
...
+ - `StateTtlConfiguration.TtlStateVisibility.ReturnExpiredIfNotCleanedUp` - returned if still available.
+ 
+ In case of `NeverReturnExpired`, the expired state behaves as if it does not exist any more, 
+ even if it has yet to be removed. The option can be useful for the use cases 
+ where data has to become unavailable for read access strictly after TTL, 
+ e.g. application working with privacy sensitive data.
+ 
+Another option `ReturnExpiredIfNotCleanedUp` allows to return the expired state before its cleanup.
+
+TTL functionality can be enabled in the descriptor of any type of state:
```
Comment (twalthr) [inline docs/dev/stream/state/state.md]: We could move this example to the beginning?

File: docs/dev/stream/state/state.md [hunk 18]
```diff
@@ -266,6 +266,101 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
...
+stateDescriptor.enableTimeToLive(ttlConfig)
+{% endhighlight %}
+</div>
+</div>
+
+**Notes:** 
+
+- The state backends store the timestamp of last modification along with the user value, 
+which means that enabling this feature increases consumption of state storage. 
+Heap state backend stores an additional java object with a reference to the user state object
```
Comment (twalthr) [inline docs/dev/stream/state/state.md]: `Java`

File: docs/dev/stream/state/state.md [hunk 19]
```diff
@@ -266,6 +266,101 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
...
+{% endhighlight %}
+</div>
+</div>
+
+**Notes:** 
+
+- The state backends store the timestamp of last modification along with the user value, 
+which means that enabling this feature increases consumption of state storage. 
+Heap state backend stores an additional java object with a reference to the user state object 
+and a primitive long in memory. RocksDB state backend adds 8 bytes per stored value, list entry or map entry.
```
Comment (twalthr) [inline docs/dev/stream/state/state.md]: `long value`

Comment (twalthr) [inline docs/dev/stream/state/state.md]: `The RocksDB state backend`

File: docs/dev/stream/state/state.md [hunk 20]
```diff
@@ -266,6 +266,101 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
...
+
+- Currently expired values are only removed when they are read out explicitly,
```
Comment (twalthr) [inline docs/dev/stream/state/state.md]: We should mentione this more explicit at the beginning. This is too hidden in my opinion. Use `<span class="label label-danger">Attention</span>`.

File: docs/dev/stream/state/state.md [hunk 21]
```diff
@@ -266,6 +266,101 @@ a `ValueState`. Once the count reaches 2 it will emit the average and clear the
...
+it won't be removed, possibly leading to ever growing state. 
+This should change in future releases. 
+Additional strategies should be added that clean up expired state automatically in the background.
+
+- Only TTLs in reference to *processing time* are currently supported.
+
+- Trying to restore state, which was previously configured without TTL, using TTL enabled descriptor or vice versa
+will lead to compatibility failure and `StateMigrationException`.
+
+- The TTL configuration is not part of check- or savepoint
```
Comment (twalthr) [inline docs/dev/stream/state/state.md]: `savepoints`

File: docs/dev/table/index.md [hunk 22]
```diff
@@ -79,4 +79,4 @@ Where to go next?
 * [Table Sources & Sinks]({{ site.baseurl }}/dev/table/sourceSinks.html): Reading tables from and emitting tables to external storage systems.
 * [User-Defined Functions]({{ site.baseurl }}/dev/table/udfs.html): Definition and usage of user-defined functions.
 
-{% top %}
\ No newline at end of file
+{% top %}
```
Comment (twalthr) [inline docs/dev/table/index.md]: Why did you remove that?

**Comment by azagrebin** [comment]
Thank you for the review @twalthr, I updated the PR.

**Comment by twalthr** [comment]
Thank you @azagrebin. I will go through the text a last time and merge this...
